File: client/java/src/main/java/io/be/mesh/system/MeshNetwork.java
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */
package io.be.mesh.system;

import io.be.mesh.macro.SPI;
import io.be.mesh.mpc.Mesh;
import io.be.mesh.mpc.ServiceProxy;
import io.be.mesh.mpc.Types;
import io.be.mesh.prsim.Context;
import io.be.mesh.prsim.Network;
import io.be.mesh.tool.Once;
import io.be.mesh.tool.Tool;
import io.be.mesh.struct.*;

import java.lang.reflect.UndeclaredThrowableException;
import java.util.List;

/**
 * @author coyzeng@gmail.com
 */
@SPI("mesh")
public class MeshNetwork implements Network {

    private final Network environs = ServiceProxy.proxy(Network.class);
    private final Once<Environ> cache = new Once<>();
    private final Context.Key<Environ> environKey = new Context.Key<>("mesh.environ", Types.of(Environ.class));

    @Override
    public Environ getEnviron() {
        try {
            return Mesh.contextSafe(() -> {
                if (cache.isPresentWithoutGet()) {
                    return cache.get();
                }
                if (Tool.required(Mesh.context().getAttribute(environKey))) {
                    return Mesh.context().getAttribute(environKey);
                }
                Environ environ = new Environ();
                environ.setNodeId("LX0000000000000");
                environ.setInstId("JG0000000000000000");
                Mesh.context().setAttribute(environKey, environ);
                return cache.get(environs::getEnviron);
            });
        } catch (RuntimeException | Error e) {
            throw e;
        } catch (Throwable e) {
            throw new UndeclaredThrowableException(e);
        }
    }

    @Override
    public boolean accessible(Route route) {
        return environs.accessible(route);
    }

    @Override
    public void refresh(List<Route> routes) {
        environs.refresh(routes);
    }

    @Override
    public Route getRoute(String nodeId) {
        return environs.getRoute(nodeId);
    }

    @Override
    public List<Route> getRoutes() {
        return environs.getRoutes();
    }

    @Override
    public List<Route> getDomains() {
        return environs.getDomains();
    }

    @Override
    public void putDomains(List<Route> domains) {
        environs.putDomains(domains);
    }

    @Override
    public void weave(Route route) {
        environs.weave(route);
    }

    @Override
    public void ack(Route route) {
        environs.ack(route);
    }

    @Override
    public void disable(String nodeId) {
        environs.disable(nodeId);
    }

    @Override
    public void enable(String nodeId) {
        environs.enable(nodeId);
    }

    @Override
    public Page<List<Route>> index(Paging index) {
        return environs.index(index);
    }

    @Override
    public Versions version(String nodeId) {
        return environs.version(nodeId);
    }

    @Override
    public Page<List<Institution>> instx(Paging index) {
        return environs.instx(index);
    }

    @Override
    public void instr(List<Institution> institutions) {
        environs.instr(institutions);
    }

    @Override
    public void ally(List<String> nodeIds) {
        environs.ally(nodeIds);
    }

    @Override
    public void disband(List<String> nodeIds) {
        environs.disband(nodeIds);
    }

    @Override
    public boolean asserts(String feature, List<String> nodeIds) {
        return environs.asserts(feature, nodeIds);
    }

}

File: client/python/mesh/system/mesh_network.py
#
# Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
# TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
#
#
from typing import List

from mesh.context import MeshKey
from mesh.kinds import Route, Environ, Versions, Paging, Page, Institution
from mesh.macro import spi
from mesh.mpc import ServiceProxy, Mesh
from mesh.prsim import Network


@spi("mesh")
class MeshNetwork(Network):

    def __init__(self):
        self.proxy = ServiceProxy.default_proxy(Network)
        self.environ: Environ = None
        self.environ_key = MeshKey("mesh-environ", Environ)

    def get_environ(self) -> Environ:
        if self.environ:
            return self.environ
        if Mesh.context().get_attribute(self.environ_key):
            return Mesh.context().get_attribute(self.environ_key)
        environ = Environ()
        environ.node_id = "LX0000000000000"
        environ.inst_id = "JG0000000000000000"
        Mesh.context().set_attribute(self.environ_key, environ)

        # self.environ = Mesh.context_safe(lambda: self.get_environ_safe())
        # return self.environ
        return environ

    def get_environ_safe(self) -> Environ:
        Mesh.context().get_principals().clear()
        return self.proxy.get_environ()

    def accessible(self, route: Route) -> bool:
        return self.proxy.accessible(route)

    def refresh(self, routes: List[Route]):
        return self.proxy.refresh(routes)

    def get_route(self, node_id: str) -> Route:
        return self.proxy.get_route(node_id)

    def get_routes(self) -> List[Route]:
        return self.proxy.get_routes()

    def get_domains(self) -> List[Route]:
        return self.proxy.get_domains()

    def put_domains(self, domains: List[Route]):
        return self.proxy.put_domains(domains)

    def weave(self, route: Route) -> None:
        return self.proxy.weave(route)

    def ack(self, route: Route) -> None:
        return self.proxy.ack(route)

    def disable(self, node_id: str) -> None:
        return self.proxy.disable(node_id)

    def enable(self, node_id: str) -> None:
        return self.proxy.enable(node_id)

    def index(self, index: Paging) -> Page[Route]:
        return self.proxy.index(index)

    def version(self, node_id: str) -> Versions:
        return self.proxy.version(node_id)

    def instx(self, index: Paging) -> Page[Institution]:
        return self.proxy.instx(index)

    def instr(self, institutions: List[Institution]) -> str:
        return self.proxy.instr(institutions)

    def ally(self, node_ids: List[str]) -> None:
        self.proxy.ally(node_ids)

    def disband(self, node_ids: List[str]) -> None:
        self.proxy.disband(node_ids)
